sklearn/tests/test_base.py

```diff
@@ -314,48 +312,131 @@ def transform(self, X, y=None):
-
-    # first, check no warning when in the same version:
+def test_pickle_version_warning_is_issued_upon_different_version():
     iris = datasets.load_iris()
-    tree = DecisionTreeClassifier().fit(iris.data, iris.target)
-    tree_pickle = pickle.dumps(tree)
-    assert_true(b"version" in tree_pickle)
-    assert_no_warnings(pickle.loads, tree_pickle)
-
-    # check that warning is raised on different version
     tree = TreeBadVersion().fit(iris.data, iris.target)
     tree_pickle_other = pickle.dumps(tree)
     message = ("Trying to unpickle estimator TreeBadVersion from "
-               "version {0} when using version {1}. This might lead to "
-               "breaking code or invalid results. "
-               "Use at your own risk.".format("something",
-                                              sklearn.__version__))
+               "version something when using version {version}. This might "
+               "lead to breaking code or invalid results. "
+               "Use at your own risk.".format(version=sklearn.__version__))
     assert_warns_message(UserWarning, message, pickle.loads, tree_pickle_other)
 
-    # check that not including any version also works:
+
+class TreeNoVersion(DecisionTreeClassifier):
+    def __getstate__(self):
+        return self.__dict__
+
+
+def test_pickle_version_warning_is_issued_when_no_version_info_in_pickle():
+    iris = datasets.load_iris()
     # TreeNoVersion has no getstate, like pre-0.18
     tree = TreeNoVersion().fit(iris.data, iris.target)
 
     tree_pickle_noversion = pickle.dumps(tree)
     assert_false(b"version" in tree_pickle_noversion)
-    message = message.replace("something", "pre-0.18")
-    message = message.replace("TreeBadVersion", "TreeNoVersion")
+    message = ("Trying to unpickle estimator TreeNoVersion from "
+               "version pre-0.18 when using version {version}. This might "
+               "lead to breaking code or invalid results. "
+               "Use at your own risk.".format(version=sklearn.__version__))
     # check we got the warning about using pre-0.18 pickle
     assert_warns_message(UserWarning, message, pickle.loads,
                          tree_pickle_noversion)
 
-    # check that no warning is raised for external estimators
+
+def test_pickle_version_no_warning_is_issued_with_non_sklearn_estimator():
+    iris = datasets.load_iris()
+    tree = TreeNoVersion().fit(iris.data, iris.target)
+    tree_pickle_noversion = pickle.dumps(tree)
     TreeNoVersion.__module__ = "notsklearn"
     assert_no_warnings(pickle.loads, tree_pickle_noversion)
+
+
+class DontPickleCacheMixin(object):
+    def __getstate__(self):
+        data = self.__dict__.copy()
+        data["_cache"] = None
+        return data
+
+
+class MultiInheritanceEstimator(BaseEstimator, DontPickleCacheMixin):
+    def __init__(self, b=5):
+        self.b = b
+        self._cache = None
+
+    @property
+    def cache(self):
```

Comment:
I still don't think this makes for the test being easily read, and would rather something more explicit than a property with side effects. Even if the test did `estimator._cache = "some_value"` directly it would be better.